cmd/gdeploy/commands/provider/add.go

```diff
@@ -27,69 +30,92 @@ var addCommand = cli.Command{
 		}
 
 		r := providerregistry.Registry()
-		p := survey.Select{
-			Message: "What are you trying to grant access to?",
-			Options: r.CLIOptions(),
+
+		uses := c.String("uses")
+		var provider *providerregistry.RegisteredProvider
+
+		if uses == "" {
+			p := survey.Select{
+				Message: "What are you trying to grant access to?",
+				Options: r.CLIOptions(),
+			}
+			var chosen string
+			err = survey.AskOne(&p, &chosen)
+			if err != nil {
+				return err
+			}
+			uses, provider, err = r.FromCLIOption(chosen)
+			if err != nil {
+				return err
+			}
+			clio.Info("Follow the documentation for setting up the %s provider here: https://docs.commonfate.io/granted-approvals/providers/%s", provider.Description, provider.DefaultID)
+		} else {
+			provider, err = r.LookupByUses(uses)
+			if err != nil {
+				return err
+			}
 		}
-		var chosen string
-		err = survey.AskOne(&p, &chosen)
-		if err != nil {
-			return err
+
+		id := c.String("id")
+		if id == "" {
+			id = dc.GetIDForNewProvider(provider.DefaultID)
 		}
-		var provider providerregistry.RegisteredProvider
-		uses, provider, err := r.FromCLIOption(chosen)
-		if err != nil {
-			return err
+
+		if _, ok := dc.Deployment.Parameters.ProviderConfiguration[id]; ok {
+			return fmt.Errorf("provider %s already exists in %s", id, f)
 		}
 
-		clio.Info("Follow the documentation for setting up the %s provider here: https://docs.commonfate.io/granted-approvals/providers/%s", provider.Description, provider.DefaultID)
+		with := map[string]string{}
+
+		withArgs := c.StringSlice("with")
 
-		var id string
-		err = survey.AskOne(&survey.Input{
-			Message: "The ID for the provider",
-			Default: provider.DefaultID,
-		}, &id, survey.WithValidator(func(ans interface{}) error {
-			str, ok := ans.(string)
-			if !ok {
-				return errors.New("couldn't validate non-string answer")
+		if len(withArgs) == 0 {
+			var pcfg gconfig.Config
+			// set up the config for the specific provider by prompting the user.
+			if configer, ok := provider.Provider.(gconfig.Configer); ok {
+				pcfg = configer.Config()
+				for _, v := range pcfg {
+					err := deploy.CLIPrompt(v)
+					if err != nil {
+						return err
+					}
+				}
 			}
-			if dc.Deployment.Parameters.ProviderConfiguration == nil {
-				return nil
+
+			err = deploy.RunConfigTest(ctx, provider.Provider)
```

Comment:
Leaving a note here that I think we will no longer be able to run these tests directly (and probably shouldn't)
When we use the assume role paradigm.

This is because the role used in the test should only have a trust relationship with the access handler execution roles.
So config tests should actually fail when the user is using a deployment role.

Because you have added an API to run these tests in the access handler, but this is not available to the CLI user either directly or via the approvals api due to auth.
We could potentially invoke the Access handler directly with a crafted request that mocks the API gateway.
Though it doesn't sound great. 

It's **critical** that we test any role relations as part of the test suite for a provider.

One solution is that tests only run via the web UI.
We can still using the cli prompt for dev/power user, but tests don't run in this workflow.

What do you think?

Reply:
Good point. Agreed that tests will now need to run server-side moving forwards. 

One possibility here is to introduce a Deployment API which is authenticated via AWS IAM, and has very limited functionality. The API could provide the testing endpoint which would work similarly to the one used in the web UI flow. The CloudFormation stack could specify allowed roles which can access this API, and when running “gdeploy init” we’d set this parameter to be the user’s current role. 

What do you think of this?

Another alternative is that the providers could be represented as CloudFormation CRDs and then create/update operations run the tests. Not sure if this is the best approach though, as we’ll want to continuously check provider connections even if the provider itself isn’t being updated.

Reply:
There a a few considerations here.

For non AWS configurables, including Identity providers for okta, azure etc that use the gconfig framework.
Testing can still happen in the CLI. (Though should we also be considering shifting these into the frontend setup workflow?)

For other non aws providers, the cli testing will work.
Its only the AWS providers where testing wont work without the assumed role.

I think we should add a condition to the tests so they only run in the cli if they are able to, and if they require the assumed role. A warning is returned instead.

If the user has gone the UI configuration route, they will have had the testing work for them.

So we can accept the poorer UX for now when using the CLI directly, which we expect our users wont need to do?

Reply:
Long term I think doing all testing server side is preferable as it will generate logs for testing and it means that all requests to integrated services are coming from a team's cloud environment rather than a development laptop.

I think your approach flagged is good if you can make a ticket for implementing the condition and warning that would be fantastic. Currently there is no validation at all if you use the UI flow